Lombiq.Tests.UI/Helpers/CloudflareHelper.cs

```diff
@@ -0,0 +1,218 @@
+using Refit;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Lombiq.Tests.UI.Helpers;
+
+internal static class CloudflareHelper
+{
+    private static readonly SemaphoreSlim _semaphore = new(1, 1);
+    private static int _referenceCount;
+
+    private static string _currentIp;
+    private static string _ipAccessRuleId;
+    private static ICloudflareApi _cloudflareApi;
+
+    public static async Task ExecuteWrappedInIpAccessRuleManagementAsync(
+        Func<Task> testAsync,
+        string cloudflareAccountId,
+        string cloudflareApiToken)
+    {
+        await _semaphore.WaitAsync();
+        Interlocked.Increment(ref _referenceCount);
+
+        Debug.WriteLine("Current reference count at the start of the test: {0}.", _referenceCount);
+
+        try
+        {
+            _cloudflareApi ??= RestService.For<ICloudflareApi>("https://api.cloudflare.com/client/v4", new RefitSettings
+            {
+                AuthorizationHeaderValueGetter = (_, _) => Task.FromResult(cloudflareApiToken),
+            });
+
+            _currentIp ??= _cloudflareApi != null ? await GetPublicIpAsync() : string.Empty;
+
+            if (_ipAccessRuleId == null)
+            {
+                Debug.WriteLine("Creating an IP Access Rule for the IP {0}.", (object)_currentIp);
+
+                // Delete any pre-existing rules for the current IP first.
```

Comment:
We can't further, because yes, there is: the first deletion is about deleting any rules that are there by mistake. There shouldn't be any, if everything worked well, but in case a previous test failed in some way that prevented its rule deletion, then this is a safety net. If we don't do this, then the rule creation will fail.

And the second deletion is about the normal process, cleaning up after the test. This is what should normally happen, but there's no guarantee it'll succeed (like if the CF API is down for an hour).